tokio-util/src/io/inspect.rs

```diff
@@ -0,0 +1,115 @@
+use futures_core::ready;
+use pin_project_lite::pin_project;
+use std::{
+    io::{IoSlice, Result},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+pin_project! {
+    /// An adapter that lets you inspect the data that's being read.
+    ///
+    /// This is useful for things like hashing data as it's read in.
+    pub struct InspectReader<R: AsyncRead, F: FnMut(&[u8])> {
```

Comment:
This is unnecessary.
```suggestion
    pub struct InspectReader<R, F> {
```

Reply:
For anyone else reading this pull request, I found [C-STRUCT-BOUNDS](https://rust-lang.github.io/api-guidelines/future-proofing.html#data-structures-do-not-duplicate-derived-trait-bounds-c-struct-bounds) that, together with [the issue that introduced it](https://github.com/rust-lang/api-guidelines/issues/6) explains why trait bounds should be scoped as narrowly as possible without making error messages worse.

Basically, this is a compat hazard - changing a trait bound causes all consumers of the `struct` to potentially need changes (since they may not have concrete types for `R` and `F`, and thus need to push trait bounds out themselves). Similar applies to `into_inner` - changing the trait bounds on a method means that callers of `into_inner` may need changing if they weren't previously aware of the trait bounds on `into_inner` (which is obviously true if we just change the bound on `F`, as that's irrelevant to `into_inner` callers).

The exception for `new` is because failing to put a trait bound on `new` would just cause this to fail to implement `AsyncRead`, and the resulting error message will not be as clear. In contrast, by putting the bounds on `new` specifically, we get a clear error message at time of construction, not time of use. In this case, the compat hazard can be avoided by simply adding an extra constructor with different bounds if the current bounds remain acceptable (e.g. we can add `impl AsyncBufRead<R, F> where R: AsyncBufRead` without needing to change the bounds on `new`, and if we want a constructor which requires `R: AsyncBufRead`, we add that as a second constructor).

Reply:
It isn't really about compat hazards. It's about making the struct more convenient to use. Generic code that uses this type needs to list the `R: AsyncRead` and `F: FnMut(&[u8])` bounds too, but if we don't require them on the struct, then code that just moves the type around (or just calls `into_inner`) does not need the bounds.